### ISSUE TITLE
CommitException: vlans target is not set

### ISSUE THREAD

**Issue by nachtmaar** [svinota/pyroute2#304]
Hi, I get the error "CommitException: vlans target is not set" during a commit.

  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 906, in commit
    self.commit(transactions=snapshots, phase=2)
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 902, in commit
    commit_mask=phase)
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/interface.py", line 905, in commit
    raise error
CommitException: vlans target is not set

Any Ideas whats wrong here?

**Comment by nachtmaar** [comment]
The version is the latest github master version.
I only setup and configure bridges (ageing_time time) and add tap devices to it.

**Comment by nachtmaar** [comment]
If I shall try another version, let me know.

**Comment by svinota** [comment]
Pls give the example that causes the error — to reproduce it. Thanks.

**Comment by nachtmaar** [comment]
I do not have a reproducible example. The commands are generated dynamically to create a certain topology. But here are the classes which mimic a bridge and a connection.

Hope that helps. I can reproduce the topology if you want. But is there a way to reconstruct that pyroute2 internally did? E.g. a description of the transaction which shall be committed? Or some log file?


At the end, the topology is created with:
```python
get_ipdb_singleton().commit()
```

```python
class BridgePyroute2(Bridge):

    '''
    Attributes
    ----------
    id : str
        Name of the bridge.
    bridge: pyroute2.ipdb.interface.Interface
    '''
    def _start(self, bridge_dev_name=None, switch=False):
        self.bridge_dev_name = bridge_dev_name

        try:
            self.bridge = get_ipdb_singleton().create(kind='bridge', ifname=self.bridge_dev_name)
            # See Also: https://github.com/svinota/pyroute2/issues/254
            # use pyroute2 functional style
            (self.bridge.set_br_ageing_time(0)
             .up()
             )
        except (NetlinkError, CreateException) as e:
            raise NetworkBackendStartError("Could not create the bridge with name '%s' in hub mode!" % self.bridge_dev_name, caused_by=e)

    def add_if(self, _if_name, if_up = True):
        try:
            tap_dev = get_ipdb_singleton().interfaces[_if_name]

            if if_up:
                tap_dev.up()
            self.bridge.up()
            self.bridge.add_port(tap_dev['index'])

        except (AttributeError, KeyError) as e:
            raise NetworkBackendBridgedBridgeError("""Could not add interface '%s' to bridge '%s'
            Bridge dump:
            %s
            Interface dump:
            %s
            Used tap devices:
            %s
            """ % (_if_name, self, check_output(["brctl", "show"]), pformat(Bridge.get_interfaces()),
                   pformat(singletons.network_backend._tap_id_mapping)), caused_by=e)
```


```python
class ConnectionPyroute2(Connection):

    def raise_lookup_device_error(self, tap_dev_name, caused_by=None):
        raise NetworkBackendConnectionError("Device '%s' not found!" % tap_dev_name, caused_by=caused_by)

    def lookup_device(self, tap_dev_name):
        try:
            idx_tap_x = get_ipdb_singleton().interfaces[tap_dev_name]
            return idx_tap_x

        except KeyError as e:
            raise self.raise_lookup_device_error(tap_dev_name, caused_by=e)

    def _tap_link_up(self, tap_x, tap_y, up=True):

        dev = self.lookup_device(tap_x)
        if up:
            dev.up()
        else:
            dev.down()
```

**Comment by svinota** [comment]
The `commit()` call should attach `.transaction` attribute to an exception it throws — usually. Also, there should be `.traceback` attribute, though it not always helps, and particularly not here as the error seems to occur in the rollback also.

And if there was an error during rollback, also `.cause` is attached (may be a bad name?)

And it looks like the time came to rewrite the `commit()` error logging :|

**Comment by nachtmaar** [comment]
Mhh, I cannot access one of these attributes.

The exception that is raised is a RuntimeError.

```python
AttributeError: 'exceptions.RuntimeError' object has no attribute 'transaction'
```

By the way, I think your loggers are not setup correctly.
I get the following error:

```
No handlers could be found for logger "pyroute2.ipdb.main"
```

**Comment by nachtmaar** [comment]
Do you need more information? Is there any other thing I can do?

**Comment by svinota** [comment]
> No handlers could be found for logger "pyroute2.ipdb.main"

By default, the library doesn't setup any logger not to conflict with the user's logging.

> Do you need more information? Is there any other thing I can do?

Probably, not yet. Give me some time to run stress tests — maybe it is a race condition. It is strange that the error reported is about vlan filtering, though you said you set up only bridging on taps.

Thanks for the reporting.

**Comment by svinota** [comment]
I changed the transaction debugging API, so now the error info should be available as `error.debug` dictionary:
```
debug = {'traceback': …,
         'transaction': …,
         'next_stage': …}
```

The `traceback` field is the formatted traceback from the `commit()` try/except statement. The `transaction` is self-clear. The `next_stage` field contains the references to the nested `commit()` stage (like a transaction rollback):
```
# `e` contains the exception
>>> print(''.join(e.debug['traceback']))
Traceback (most recent call last):
  File "pyroute2/ipdb/interface.py", line 727, in commit
    run(nl.link, 'set', **request)
  File "pyroute2/ipdb/interface.py", line 455, in _run
    raise error
NetlinkError: (22, 'Invalid argument')

>>> print(e.debug['transaction'])
{'txqlen': 1000,
 'ipdb_scope': 'system',
 'index': 1610,
 'operstate': 'DOWN',
 'num_tx_queues': 1,
 'group': 0,
 'carrier_changes': 1,
 'ipaddr': [],
 'ifname': 'bala',
 'promiscuity': 0,
 'linkmode': 0,
 'kind': 'veth',
 'broadcast': 'ff:ff:ff:ff:ff:ff',
 'link': 1609,
 'address': '3a:e9:38:fb:8f:80',
 'peer': 'dala',
 'vlans': [],
 'ipdb_priority': 0,
 'gso_max_segs': 65535,
 'gso_max_size': 65536,
 'qdisc': 'noop',
 'mtu': 67,
 'num_rx_queues': 1,
 'carrier': 0,
 'flags': 4098,
 'proto_down': 0,
 'ports': []}
```

(In the example above the **mtu** field is a reason to fail)

**Comment by nachtmaar** [comment]
Thanks for the code update.
I repeated the experiment some times and got two different exceptions.

Error 1:
```python
CommitException: ipaddr target is not set

{'debug': {'next_stage': None,
  'traceback': 'Traceback (most recent call last):\n  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/interface.py", line 828, in commit\n    raise CommitException(\'ipaddr target is not set\')\nCommitException: ipaddr target is not set\n',
  'transaction': {'address': 'de:d3:ba:63:cc:9a',
   'broadcast': 'ff:ff:ff:ff:ff:ff',
   'carrier': 1,
   'carrier_changes': 0,
   'flags': 4098,
   'group': 0,
   'ifname': 'tap_00120_2',
   'index': 511284,
   'ipaddr': [],
   'ipdb_priority': 0,
   'ipdb_scope': 'system',
   'kind': 'tun',
   'linkmode': 0,
   'mtu': 1500,
   'num_rx_queues': 1,
   'num_tx_queues': 1,
   'operstate': 'DOWN',
   'ports': [],
   'promiscuity': 0,
   'qdisc': 'noop',
   'txqlen': 500,
   'vlans': []}}}
```
Error 2:
```python
KeyError: 'vlans'
```

**Comment by svinota** [comment]
Yep, thanks.

**Comment by nachtmaar** [comment]
If you have new code that shall be tested, tell me :)
Its easy to rerun the test.

**Comment by svinota** [comment]
Pls try meanwhile to increase the commit timeout — in the case the system is overloaded. Just to check it is not the case. Thanks.
```
from pyroute2.ipdb import transactional
transactional.SYNC_TIMEOUT = 15
```

**Comment by nachtmaar** [comment]
Unfortunately, it does not help Even with 150 it does not.

**Comment by svinota** [comment]
At least, we can exclude it now from the possible causes.

**Comment by nachtmaar** [comment]
Any news? 

Pyroute2 ist part of my master thesis and I want to benchmark it against iproute2 in terms of time it takes until a network topology has been created.

Would be nice, if we can do it until the deadline :)

**Comment by svinota** [comment]
@nachtmaar that's bad, I didn't know about the deadline.

I'm tryin to add tap as ports to bridges + set `br_ageing_time`, but in neither combination can I reproduce the error :| The thing that freaks me out is that it reports about the `vlans` target on the bridge, though there should be no vlan filter setup.

**Comment by svinota** [comment]
May it be a race condition between two different threads trying to set up the same interface?

**Comment by nachtmaar** [comment]
Mhh, I provided a lock around `get_ipdb_singleton().commit()`.

I repeated three times the experiment and now the only exception I still get is the one with the ` ipaddr target is not set`.

Maybe we can focus on this error?

Is there an action log I can view?

Like:
```bash
ip l s dev eth0 up
... more commands
```

The debug dictionary:
```python
{
    "debug": {
        "traceback": "Traceback (most recent call last):\n  File \"/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/interface.py\", line 828, in commit\n    raise CommitException('ipaddr target is not set')\nCommitException: ipaddr target is not set\n",
        "next_stage": null,
        "transaction": {
            "txqlen": 500,
            "ipdb_scope": "system",
            "index": 669084,
            "operstate": "DOWN",
            "num_tx_queues": 1,
            "group": 0,
            "carrier_changes": 0,
            "ipaddr": [],
            "ifname": "tap_00038_1",
            "promiscuity": 0,
            "linkmode": 0,
            "broadcast": "ff:ff:ff:ff:ff:ff",
            "address": "f6:e7:1f:95:f5:c1",
            "vlans": [],
            "ipdb_priority": 0,
            "kind": "tun",
            "qdisc": "noop",
            "mtu": 1500,
            "num_rx_queues": 1,
            "carrier": 1,
            "flags": 4098,
            "ports": []
        }
    }
}
```

**Comment by svinota** [comment]
Look, it is the `ipaddr` target. In the transaction `ipaddr` is empty, `[]`. It means that IPDB expects the interface to have no addresses, but the interface still has some addresses.

It may happen e.g. if ip addresses are added in the very moment the `commit()` operates on them, as the `ipaddr` field in the interface (not transaction) object is continuously updated in the runtime from the system, and normally it should not cause an error (but there is still some possibility, since we can not lock the Linux kernel, thus races are still possible with other programs).

So again it seems to be a race, but a different one. May it be that the NetworkManager runs on the same system?

I'll try to provide tomorrow a branch with the full debug in the IPDB, but not sure how bad it will affect the performance, and there is a possibility — if it is a race condition — that the debug logging by itself will shadow the error by slowing down the core.

**Comment by svinota** [comment]
IPDB works on ip addresses in the `commit()` after the interface `flags` are set. The flags in the transaction are `4098` — the interface should be **down** after `commit()`

What could happen:
1. The interface is set **up** (thus implicitly bringing up a link-local IPv6 address) again after it is set down by IPDB and addresses are flushed, but before IPDB checks the `ipaddr` target (rare, but still possible case)
2. There are some addresses left while in the transaction, despite the explicit flush (can not imagine, how it may happen) — then it is a library bug.

**Comment by nachtmaar** [comment]
I think I found the problem! :)

After you mentioned the interface states, I commented out all code that brings up the tap device.
And now it works :)

Depending on the topology I want to create, a tap devices state might be set multiple times to up.

Therefore multiple times calling `dev.up()`. Is this the problem? Do I have to take core of this myself?

I though the commit() call will take core of the multiple calls and perform only one?

Btw, there is no NetworkManager running on the system. And I encountered the vlan exception again, but only with the outcommented code!

**Comment by svinota** [comment]
> Therefore multiple times calling dev.up() (sequentially). Is this the problem?

It must not be a problem. When you change an interface state, the change is reflected not in the interface object itself, but instead it is transparently proxied into a «transaction» object. The `commit()` call should apply only the difference between the transaction and the interface.

The `up()` call should only set the first bit in the `flags` — you can do it manually also. Multiple `up()` calls don't make any difference, 1 is always 1.

That's cool that it starts to work. Pls let me know, if there will be any problems. And the debug code will be ready soon anyways, just in case.

**Comment by nachtmaar** [comment]
I am still having problems. Is the debug code ready?

I modified my code to write a history of pyroute2 calls and interestingly now I get an additional exception before the `ipaddr target` exception occurrs.

I appended the ipyroute2 command log. Note that the tap devices are created by Qemu beforehand.

```python
ERROR pyroute2.ipdb.main serve_forever: Restarting IPDB instance after error:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 932, in serve_forever
    messages = self.mnl.get()
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/netlink/nlsocket.py", line 701, in get
    data = self.recv_ft(bufsize)
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/netlink/nlsocket.py", line 934, in recv_plugin
    raise data_in
error: [Errno 105] No buffer space available

ERROR pyroute2.ipdb.main serve_forever: Error restarting DB:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 947, in serve_forever
    self.initdb()
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 647, in initdb
    self.release(complete=False)
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 821, in release
    self._mthread.join()
  File "/usr/lib/python2.7/threading.py", line 940, in join
    raise RuntimeError("cannot join current thread")
RuntimeError: cannot join current thread

CRITICAL MiniWorld 2016-11-08 10:19:31,651 NetworkBackendBridged: {'debug': {'traceback': 'Traceback (most recent call last):\n  File "/usr/local/lib/python2.7/dist-packages/pyroute
2/ipdb/interface.py", line 828, in commit\n    raise CommitException(\'ipaddr target is not set\')\nCommitException: ipaddr target is not set\n', 'next_stage': None, 'transaction':
{'txqlen': 500, 'ipdb_scope': 'system', 'index': 684604, 'operstate': 'DOWN', 'num_tx_queues': 1, 'group': 0, 'carrier_changes': 0, 'ipaddr': [], 'ifname': 'tap_00120_2', 'promiscui
ty': 0, 'linkmode': 0, 'broadcast': 'ff:ff:ff:ff:ff:ff', 'address': 'a2:0b:09:6e:57:e2', 'vlans': [], 'ipdb_priority': 0, 'kind': 'tun', 'qdisc': 'noop', 'mtu': 1500, 'num_rx_queues
': 1, 'carrier': 1, 'flags': 4098, 'ports': []}}}
CRITICAL MiniWorld 2016-11-08 10:19:31,652 NetworkBackendBridged: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/interface.py", line 828, in commit
    raise CommitException('ipaddr target is not set')
CommitException: ipaddr target is not set
```

[pyroute2_history.txt](https://github.com/svinota/pyroute2/files/577855/pyroute2_history.txt)

**Comment by nachtmaar** [comment]
Moreover,the problem does not occur if all the tap.up() calls are commented out

**Comment by svinota** [comment]
Wow. Looks like neither the ipaddr nor the state is an issue, but it is the IPDB restart that screws all the things.

The IPDB doesn't restart by itself, it is a netlink socket error that triggers the restart. The socket error may come if the netlink client doesn't manage to handle all the incoming messages.

The IPDB normally runs the netlink messages fetching code as a separate thread. The code doesn't parse the messages, only buffers them to the parser (see also `IPDB._nl_async`). If there is a message burst, you'll get in the logs (if the logging is set up): https://github.com/svinota/pyroute2/blob/master/pyroute2/netlink/nlsocket.py#L43

If you get the ENOBUF error (https://github.com/svinota/pyroute2/blob/master/pyroute2/netlink/nlsocket.py#L23) it may mean that the IPDB runs on the top of `NetNS` (that doesn't respect `_nl_async` 'cause of the architectural issues), not `IPRoute` — is it the case? Or maybe you manually set `_nl_async` to `False` somehow?

If it is the case, maybe it is better to spawn normal IPDB _inside_ netns, like
1. spawn a child
2. setns()
3. instantiate IPDB() w/o any additional parameters

**Comment by svinota** [comment]
@nachtmaar can we meet on the IRC channel #pyroute2 on freenode, or even arrange a video meeting in GTalk / Hangouts? Maybe it would be easer.

**Comment by nachtmaar** [comment]
Sure, I joined the chat on IRC ;)

**Comment by svinota** [comment]
So the issue is that the monitoring socket of the IPDB doesn't manage to handle all the messages arriving. I see two ways to solve that.

### 1. Exclude the monitoring socket at all

In that case we have to use an `IPRoute` object. Thus, the sequence will be like that:
1. Build the topology cache `{'name': ('kind', int(index), …), …}`
2. Iterate the cache, creating interfaces: `IPRoute.link("add", kind=…, ifname=…, …)`
3. Create a set of interface names: `s = set(cache.keys())`
4. Call `IPRoute.get_links()`, excluding all the names from the set
5. If the set is not empty, sleep(~0.1), repeat 4.
… At this step we have all the interfaces created,  and the system state is in sync.
6. Iterate the cache, setting interface parameters
… We're done.

### 2. Improve the message monitoring routine

We can not affect other Python threads, there is no priority; so one of the solutions would be to have the monitoring in a separate _process_, not a Python thread. Such process could be even pinned to a CPU in the very same way as with DPDK or OVS. I'll try to make a proof-of-concept in one-two days.

**Comment by nachtmaar** [comment]
Ok, thanks :)

Because I have a lot of other work to do, I will try solution 1 if solution 2 does not work but only if there is enough time left at the end :)

**Comment by nachtmaar** [comment]
Nethertheless, I you want me to test new pyroute2 versions, feel free to ask ;)

**Comment by svinota** [comment]
Pls try the `async_cache` branch.

Usage: `ipdb = IPDB(nl_async="process")`

Pls be aware, that a child Python process is spawned. All the monitoring is done then via that process. The process can have a different priority and/or you can pin it to a CPU.

It is a quick-n-dirty hack, so there **are** issues. But it **may** work for you.

**Comment by nachtmaar** [comment]
Unfortunately it is not working :/

Installed from the "async_cache" branch and used ipdb = IPDB(nl_async="process")`

```python
ERROR pyroute2.ipdb.main serve_forever: Restarting IPDB instance after error:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 934, in serve_
forever
    messages = self.mnl.get()
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/netlink/nlsocket.py", line 708, in
 get
    data = self.recv_ft(bufsize)
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/netlink/nlsocket.py", line 942, in
 recv_plugin
    raise data_in
error: [Errno 105] No buffer space available

ERROR pyroute2.ipdb.main serve_forever: Error restarting DB:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 949, in serve_forever
    self.initdb()
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 649, in initdb
    self.release(complete=False)
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 823, in releas
e
    self._mthread.join()
  File "/usr/lib/python2.7/threading.py", line 940, in join
    raise RuntimeError("cannot join current thread")
RuntimeError: cannot join current thread

"/root/miniworld/MiniWorld/miniworld/model/network/backends/bridged/pyroute2/NetworkBackendBridged.py", line 95, in do_network_topology_change
    raise e
KeyError: 'vlans'
```

**Comment by nachtmaar** [comment]
Hi,

I tried to recreate the code that leads to the error.
See the attached file (had to rename it to ".txt")
Unfortunately, it creates another exception, but I guess I am creating the tap devices wrong. In my case they are created by Qemu.
But nethertheless, maybe it helps you ;)

[pyroute2_error.txt](https://github.com/svinota/pyroute2/files/580490/pyroute2_error.txt)

**Comment by svinota** [comment]
Could you please check if it really spawns a child Python process?

Thanks for the code, I'm trying it now.

**Comment by nachtmaar** [comment]
Sure. Btw, it worked a few times now :) But still sometimes not, think more than 50%.

Also note, that there are actually 2 commits().
The mgmt bridge is used for management purposes inside the simulator and created in the first commit.
Afterwards the topology is setup with the second commit()

I attached a new file.

RPC Server Process:
```python
>>> threading.enumerate()
[<ExceptionStopThread(Thread-1, started daemon 140262303074048)>, <Thread(IPDB event loop, started daemon 140261216745216)>, <ReversePythonShell(Thread-134, started 140257953556224)>, <RunLoop(Thread-68, started daemon 140262215509760)>, <_MainThread(MainThread, started 140262411499328)>, <ExceptionStopThread(Thread-67, started daemon 140257433470720)>]
```

I guess your process:
```python
>>> threading.enumerate()
[<_MainThread(MainThread, started 140262411499328)>, <Thread(QueueFeederThread, started daemon 140257433470720)>, <ReversePythonShell(Thread-68, started 140262215509760)>]
```


[pyroute2_error.txt](https://github.com/svinota/pyroute2/files/580603/pyroute2_error.txt)

**Comment by nachtmaar** [comment]
Hi, any news? Could you reproduce the behaviour?

**Comment by svinota** [comment]
I've got some other issues with the script I had to fix. But not that one. I'm continuing with the debug.

**Comment by svinota** [comment]
Meanwhile — try to prioritize the cache process. The process is seen as a child, just set the highest priority to it with `renice`.

**Comment by svinota** [comment]
Still can not reproduce this particular error. Probably I have too fast system.

Pls take a look at this patchset: https://github.com/svinota/pyroute2/pull/314

I cherry-picked the code to the `async_cache` branch. Try to minimize the netlink traffic by switching off unneeded RTNL groups.

Another step would be to commit not such huge batch of changes in one turn — but in several iterations.

**Comment by nachtmaar** [comment]
Thanks, I will try it out :)

Btw, how can I check for errors in the batch mode of IPRoute?
Maybe I will try out the IPRoute alternative :)

```python
import pyroute2
ipr = pyroute2.IPRoute()
ipb = pyroute2.IPBatch()
ipb.link("add", index=550, kind="dummy", ifname="test2")
ipr.sendto(ipb.batch, (0, 0))
```

**Comment by nachtmaar** [comment]
Can you tell me if these groups are sufficient:
```python
GROUPS = rtnl.RTNLGRP_LINK | rtnl.RTNLGRP_NOTIFY | rtnl.RTNLGRP_NEIGH
ipdb = IPDB(nl_async="process", nl_bind_groups=GROUPS)
```

I get the following error:

```python
  File "/root/miniworld/MiniWorld/miniworld/model/network/backends/bridged/pyroute2/NetworkBackendBridged.py", line 75, in do_network_topology_change
    ipdb.commit()
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 883, in commit
    in self.by_name.values() if x.local_tx.values()]
  File "/usr/local/lib/python2.7/dist-packages/pyroute2/ipdb/main.py", line 709, in __getattribute__
    return super(IPDB, self).__getattribute__(name)
AttributeError: 'IPDB' object has no attribute 'by_name'
```

Moreover, note that I finally get the burst errors you were looking for :-D

They appear if I interrupt the program after the exception with CTRL-C:

```python
DEBUG pyroute2.netlink.nlsocket get: Packet burst: delta=28 qsize=162 delay=0.01
DEBUG pyroute2.netlink.nlsocket get: Packet burst: delta=20 qsize=182 delay=0.01
DEBUG pyroute2.netlink.nlsocket get: Packet burst: delta=16 qsize=198 delay=0.01
DEBUG pyroute2.netlink.nlsocket get: Packet burst: delta=20 qsize=218 delay=0.01
```

**Comment by svinota** [comment]
> Btw, how can I check for errors in the batch mode of IPRoute?

If the `sendto()` method doesn't raise an exception, it means that the kernel accepted the messages, and there is no other way to check if the kernel implemented everything requested, than to monitor broadcasts or poll for RTNL objects for `IPRoute.get_whatever()`

> Can you tell me if these groups are sufficient:

To work with the links in IPDB, one should enable these groups:

* rtnl.RTNLGRP_LINK
* rtnl.RTNLGRP_NEIGH
* rtnl.RTNLGRP_IPV4_IFADDR
* rtnl.RTNLGRP_IPV6_IFADDR

> Moreover, note that I finally get the burst errors you were looking for :-D

Pls try to renice the netlink cache process, if it is possible (maybe even give it a realtime priority) — for me it looks like Python just works too slow on your testbed

**Comment by nachtmaar** [comment]
I have good and bad news.

The bad news are that IPDB is still not working for me. Note that I have two network switching modes: In the first, tap devices are added to a bridge and then ebtables is used to govern over connectivity of nodes.
The second uses one bridge per connection where two NICs are added to. This version is the one that does not work with IPDB, maybe because there are a lot more of NICs involved.

I used the following flags:

```python
rtnl.RTNLGRP_LINK |\
rtnl.RTNLGRP_NEIGH |\
rtnl.RTNLGRP_IPV4_IFADDR |\
rtnl.RTNLGRP_IPV4_ROUTE |\
rtnl.RTNLGRP_IPV4_MROUTE |\
rtnl.RTNLGRP_IPV6_IFADDR |\
rtnl.RTNLGRP_IPV6_ROUTE |\
rtnl.RTNLGRP_MPLS_ROUTE
```

Using less resulted in either

```python
AttributeError: 'IPDB' object has no attribute 'routes'
```

or:

```python
AttributeError: 'IPDB' object has no attribute 'by_name'
```

The IPDB instance restarts after a crash but this does not overcome the actual problems. Even renicing or setting the realtime attributes (not sure if I did it correctly) did not help. I used the `self.ipdb.mnl.async_cache.pid` as PID for renicing,


The good news: I finally switched my implementation to use IPRoute together with your described caching behaviour and the results are inline with the ones from iproute2 in batch mode :)
